### PR TITLE
Performance tweaks

### PR DESCRIPTION
**Pull request by jcamilleri13**
Proposed performance tweaks for deep gets, discussed in #41. Function falls back to previous linear search if key not found and strict mode is disabled. If strict mode is enabled (through global config) the function will throw an error.

### CODE REVIEW

File: sass/_core.scss [hunk 1]
```diff
@@ -23,6 +23,8 @@
 ///   and provides specific tools for maintaining those values
 /// @group core
 
+$_a_strict: false !default;
```
[the remark this answers is not in the document]
Comment (jcamilleri13) [reply sass/_core.scss]:
```suggestion
$strict-mode: false !default;
```
or perhaps
```suggestion
$error-on-missing-keys: false !default;
```
to make it really obvious? I'll add some inline documentation for the option for Sassdoc to pick up. Because of how the code is handling locating aliases, adding a `/` as a terminating character is actually trivial, I'll add it in tomorrow and write a test.

File: test/core/_parser.scss [hunk 2]
```diff
@@ -245,6 +256,17 @@
       'N4M3!ofthething'
     );
   }
+
+  @include it('Fails if given an alias without explicit ending when strict mode is enabled') {
```
[the remark this answers is not in the document]
Comment (jcamilleri13) [reply test/core/_parser.scss]: The token termination is easy to handle as I've mentioned above. and having an escape for the # symbol could probably easily be added with a look-behind kind of thing. Will look into the latter.

**Comment by jcamilleri13** [comment]
So I had actually implemented the single sort tweak for the non-strict mode, but the first time I had implemented it I hadn't realised that the function was being called recursively. While it's way faster than sorting for every single lookup, it's not quite a single sort of the whole flattened map. Since with the last commit the get-token only results in a linear search if the key cannot be found at all, as a sort of fallback, I wasn't sure if it was worth pre-sorting the map. I suppose it would depend on the real world usage - if a high perecentage of `get-token()` calls are made with non-existent keys, the pre-sorting is worth it, but if it's only a couple of missed hits this way might actually be faster. Not entirely sure how I'd go about testing that.

I can write the documentation, where would you suggest it goes? As a sassdoc comment for the configuration flag or should it part of the more general explanation?

File: sass/_core.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (mirisuzanne) [reply sass/_core.scss]: I like the more explicit `$error-on-missing-keys`, but don't have permissions to commit on your branch. :)

**Comment by mirisuzanne** [comment]
> if a high perecentage of `get-token()` calls are made with non-existent keys, the pre-sorting is worth it, but if it's only a couple of missed hits this way might actually be faster. Not entirely sure how I'd go about testing that.

Got it - let's leave it out for now. (The map flattening is a similar trade-off, but for usage of `->`)

> I can write the documentation, where would you suggest it goes? As a sassdoc comment for the configuration flag or should it part of the more general explanation?

I think it works as a comment on the flag. We might consider defaulting it `true` when we do a major release, and could add summary docs at that point.

File: sass/_core.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcamilleri13) [reply sass/_core.scss]: I have no idea how this feature works and I've mucked something up. Renaming it on my branch so you can pull the corrected code.

File: test/core/_parser.scss [hunk 3]
```diff
@@ -279,16 +279,16 @@
...
-  //   $error-on-missing-keys: true !global;
+  @include it('Finds an explicitly terminated alias when strict mode is enabled') {
+    $error-on-missing-keys: true !global;
 
-  //   @include assert-equal(
-  //     _a_replace($map, '#name\\ofthething'),
-  //     'N4M3!ofthething'
-  //   );
+    @include assert-equal(
+      _a_replace($map, '#name\\ofthething'),
```
Comment (mirisuzanne) [inline test/core/_parser.scss]: If I understand right, since `\` is already an escape character, users would have to use `\\`? I don't know if that's ironic, or a problem, or nice that it forces even more explicit intent…  😛 In any case, we'd want to make sure it's clear in the docs.

Comment (jcamilleri13) [reply test/core/_parser.scss]: It's actually because of how css & sass actually handle strings: [https://sass-lang.com/documentation/values/strings#escapes](https://sass-lang.com/documentation/values/strings#escapes).

Since the /character escapes any other character besides0-9 and A-F (the first point in that documentation item) writing `/#` is just parsed as`#`. You'd need to write it twice to escape itself, so to speak. So the parser reads `\\#` as `\#`. Valid point that it should be documented though.

Comment (mirisuzanne) [reply test/core/_parser.scss]: Yeah, that's what I was getting at. It's a Sass-escape character, so we have to Sass-escape it in order to generate our own escape character. A double-escape! Escaping all the way down!

Comment (mirisuzanne) [reply test/core/_parser.scss]: I don't know if there's some danger with that, but I kinda like that it means our escape is practically `\\` as far as users are concerned.

Comment (jcamilleri13) [reply test/core/_parser.scss]: I think it's fine, it means they can use unicode characters or escapes in their strings without any hassle. Maybe we should write some more tests at some point to cover any weird edge cases.

**Comment by jcamilleri13** [comment]
So I just thought about something regarding the error on missing keys. Would it make more sense to return `null`, as the standard `map-get()` does? Or should it be another option? Or is failing outright more sensible? It occurred to me where there might be cases that a user would want to carry out an alternate action if a key doesn't exist, and throwing an error doesn't allow for that.

**Comment by mirisuzanne** [comment]
Since we're adding a toggle anyway - would it be difficult for that toggle to allow several values? 

- the legacy silent fail
- return null
- return error

Is `null` ever useful for internal map references, or only for explicit `token-get()` calls?

**Comment by jcamilleri13** [comment]
I wasn't sure what you meant by that last bit, but now that I've started coding it I think I know what you're getting at. We probably shouldn't return `null` for internal references, it's akin to a silent fail. Or rather we should ensure that the `null` carries on up the chain. So if I were to request a `get-token()` for a key that maps to the following string: `'#ref1 #ref2->ref3 10px'`, and `'ref3'` doesn't exist, should it return `'value1 10px'` or `null`? I'm leaning towards the latter, because it will be impossible to know if bits of your references are wrong because of how they'll get converted into strings.

Also the tests are currently a bit slapdash, I'd like to beef them up a bit more to handle these edge cases. Should they be sectioned off into their own region for tests specifically related to the error handling? We could avoid that unsightly toggling of the error modes too.

**Comment by mirisuzanne** [comment]
Yeah, that's the issue I was thinking about. I think you're right about the best way to handle it if we're using `null` - but I'm not entirely sure that works as a universal solution: we may still want a toggle of some kind. An explicit error would be more helpful in debugging *where* things went wrong. I can see places I'd want silent failure like that (e.g. `get-token('optional') or $fallback`) - but also a lot of times that I would rather have the explicit error behavior.

I'm pretty flexible on where the tests live: some balance of:

- roughly match the code structure itself
- break things out into sensible chunks

**Comment by mirisuzanne** [comment]
@jcamilleri13 What's the status of this? I got distracted with some traveling, but I'd love to get it merged in.

**Comment by jcamilleri13** [comment]
Hey, sorry about that, literally just walked off the plane too. So the code is functioning with a 3-way toggle (silent, null, and error). Should have committed all the changes to the issue branch, but I'll double check them tomorrow. Regarding the tests, I've we're good with a three way switch I'm going to break them out into a separate file because they feel a bit ham-fisted in their current placement.

**Comment by mirisuzanne** [comment]
@jcamilleri13 That all work for me, and sounds great! 

No rush, we're all working on Open-Source Time, just didn't want to lose track of it. :)

**Comment by jcamilleri13** [comment]
So I'm (finally!) wrapping up the tests on this so we can get the merge done, I'm hoping it can help out with #48 too. I thought of an additional enhancement to the three-way toggle for missing map items – a fourth option, `warn` which would act similarly to the `null` option, whilst also throwing a warning for traceability purposes. The only issue I have here is that I don't think there's any way to write a test for a `@warn` command in sass.

**Comment by mirisuzanne** [comment]
If you look at how we handle `@error` testing, the same approach can be used for `@warn`. Requires adding `_a_warn()` function/mixin that uses `$_a_error-output-override` to toggles between an actual warning or returning the text of the warning.

**Comment by jcamilleri13** [comment]
Yeah I thought of that, the only thing is that `_a_error()` actually returns the error, which is fine because you'd expect the code to terminate on an error and have no need for an actual `@return` statement after. With a warning however you still want your code to function normally, so putting a `@return _a_warn(...)` statement in will break your intended functionality. You could return a tuple with the intended output plus the warning, but I'm not sure if the benefits outweigh the muddiness added to the codebase.

**Comment by mirisuzanne** [comment]
No, I don't think that's worth it. If we can get away with treating it like an error just for testing purposes, go for it. Otherwise I suppose we live with it for now…

**Comment by jcamilleri13** [comment]
A neater solution would perhaps be returning the warning if `$_a_error-output-override` is set to `true`, and returning the actual value if it's `false`. We wouldn't be able to test the actual return value though, and I'm worried that increases the chance of letting errors slip through. A workable but ugly solution is to use some global variable but I'm not really a fan of that at all.

**Comment by mirisuzanne** [comment]
Yeah, I think we have to live with not testing warnings…

**Comment by jcamilleri13** [comment]
Ok, warn mode added and unit tests handle most of the cases I think. I kept the same layout since the test files mirrored the existing code structure. @mirisuzanne please review at your convencience. Sorry it's taken so long and thanks for the help!

File: sass/core/_parser.scss [hunk 4]
```diff
@@ -105,6 +105,44 @@
...
+// -----
+/// Returns the next available key for parsing/replacing from the given string.
+///
+/// @access private
+///
+/// @param {string} $string -
+///   A string to extract an #alias from
+/// @return {string | null} -
+///   The extracted #alias key to replace, or null if no references exist
+@function _a_next_reference(
```
Comment (mirisuzanne) [inline sass/core/_parser.scss:118]: I think we still need tests for `_a_next_reference()`?

File: sass/core/_api.scss [hunk 5]
```diff
@@ -305,3 +305,22 @@
...
+///         and greedily replace all existing references.
+/// null:   `get-token()` will return `null` if the value is not found,
+///         mimicking sass's inbuilt `map.get()` function.
+/// warn:  `get-token()` will return `null` if the value is not found, and
+///         also throw a warning.
+/// error: `get-token()` will throw an error if the value is not found.
+///
+/// @group core-get
+/// @type string
+$handle-missing-keys: 'silent' !default;
```
Comment (mirisuzanne) [inline sass/core/_api.scss:326]: This is the only user-facing change, right? The rest is internal performance optimization?

Comment (jcamilleri13) [reply sass/core/_api.scss:326]: Yep, just that config. Everything else is under the hood, and the current default is the existing behaviour. Should we throw an error if the value isn't one of the four possible options, or just allow it to carry out the default silent failure as it does currently?

File: sass/core/_parser.scss (hunk 4 above)
Comment (jcamilleri13) [reply sass/core/_parser.scss:118]: ...I completely forgot I wrote this code. Unit tests coming right up!

File: sass/core/_api.scss (hunk 5 above)
Comment (mirisuzanne) [reply sass/core/_api.scss:326]: I suppose an error would be most clear and explicit - good call.

Comment (jcamilleri13) [reply sass/core/_api.scss:326]: Added the error, plus corresponding test. It's in the `_a_get()` function.

File: sass/core/_parser.scss (hunk 4 above)
Comment (jcamilleri13) [reply sass/core/_parser.scss:124]: Unit tests for `_a_next_reference()` added. :)

**Comment by jcamilleri13** [comment]
Hey @mirisuzanne is there anything else I need to do to prepare for the merge? I'm a tad new to this so not sure if I'm missing anything.

**Comment by jcamilleri13** [comment]
Gentle reminder about this merge please. I have an internal company project that is relying on my own hacky version and I'd like to bring it in line with the latest version of the library when possible :)

File: sass/core/_parser.scss [hunk 6]
```diff
@@ -27,6 +27,12 @@
   $do: null;
   $value: $key;
 
+  // Throw an error if the missing key flag is not configured correctly.
+  @if not index(('error', 'warn', 'null', 'silent'), $handle-missing-keys) {
```
Comment (mirisuzanne) [inline sass/core/_parser.scss:31]: This only allows `'null'` as a string value. I was expecting it to be an actual `null`. We could support both, or we should clearly document which one we mean. I'd be ok with either, but think `null` has to be valid without the quotes. (from a quick test, I think adding unquoted null to the list is all we need for a fix here)

**Review by mirisuzanne — CHANGES_REQUESTED** [review]
Hi looking great. Sorry, wild summer. I noticed this one potential confusion in the error, but otherwise happy to merge.

File: sass/core/_parser.scss (hunk 6 above)
Comment (jcamilleri13) [reply sass/core/_parser.scss:31]: Ah, yes. I assumed a string value would make more sense for configuration, as opposed to an actual null value since that tends to have secondary effects in SASS. Supporting both would a safe bet, I'll update the list to also include an actual `null` value.

Comment (jcamilleri13) [reply sass/core/_parser.scss]: Updated the configuration to also include `null`. :)
Thanks for looking into this, appreciate the time.

Comment (mirisuzanne) [reply sass/core/_parser.scss]: @jcamilleri13 I can merge and make this change myself if that's faster - just wondered if you had thoughts on it.

**Comment by mirisuzanne** [comment]
@jcamilleri13 I just released v2.2.0 with your changes. Thank you! 🎉

**Comment by jcamilleri13** [comment]
Super, thank you! 😊
